src/targets/gpu/device/scatter.cpp

```diff
@@ -15,20 +15,22 @@ argument scatter(
 {
     auto ds            = arg0.get_shape();
     auto inds          = arg1.get_shape();
+    auto upds          = arg2.get_shape();
     auto axis_dim_size = ds.lens()[axis];
-    hip_visit_all(result, arg0, inds)([&](auto output, auto data, auto s1) {
+    hip_visit_all(result, arg0, inds, upds)([&](auto output, auto data, auto inds1, auto upds1) {
         auto* output_ptr     = device_cast(output.data());
         const auto* data_ptr = device_cast(data.data());
         gs_launch(stream, ds.elements())([=](auto i) __device__ { output_ptr[i] = data_ptr[i]; });
         hip_visit_all(arg1, arg2)([&](auto indices, auto update) {
             const auto* upd_ptr     = device_cast(update.data());
             const auto* indices_ptr = device_cast(indices.data());
             gs_launch(stream, inds.elements())([=](auto i) __device__ {
-                auto out_idx    = s1.multi(i);
-                auto index      = indices_ptr[i];
-                index           = index < 0 ? index + axis_dim_size : index;
-                out_idx[axis]   = index;
-                output[out_idx] = upd_ptr[i];
+                auto out_idx    = inds1.multi(i);
```

Comment:
Just do `auto out_idx = indices.get_shape().multi(i)`.

Reply:
These changes are triggering a series of Compile-time errors.

Reply:
Whats the compile error?

Reply:
Following is the snippet : 

```
774 /home/umayadav/repo/AMDMIGraphX/src/targets/gpu/device/include/migraphx/gpu/device/shape.hpp:36:41: note: candidate function not viable: no known conversion from 'hip_array<[...], 5>' t    o 'hip_array<[...], 1U aka 1>' for 1st argument
775     MIGRAPHX_DEVICE_CONSTEXPR index_int index(hip_index x) const { return x.dot(strides); }
776                                         ^
777 /home/umayadav/repo/AMDMIGraphX/src/targets/gpu/device/include/migraphx/gpu/device/shape.hpp:38:41: note: candidate function not viable: no known conversion from 'migraphx::gpu::device:    :hip_array<unsigned int, 5>' to 'std::initializer_list<index_int>' (aka 'initializer_list<unsigned int>') for 1st argument
778     MIGRAPHX_DEVICE_CONSTEXPR index_int index(std::initializer_list<index_int> x) const
779                                         ^
780 /home/umayadav/repo/AMDMIGraphX/src/targets/gpu/device/include/migraphx/gpu/device/shape.hpp:46:41: note: candidate function not viable: no known conversion from 'migraphx::gpu::device:    :hip_array<unsigned int, 5>' to 'migraphx::gpu::device::index_int' (aka 'unsigned int') for 1st argument
781     MIGRAPHX_DEVICE_CONSTEXPR index_int index(index_int i) const

```
It happens when i do `output[out_idx] = update[upd_idx]`.   during the tensor-view of `update[upd_idx]` it calls into `index()` and then errors about no matching function.

Reply:
@umangyadav You can refer the gpu/device/where.cpp file to use the if constexpr to see whether it can be fixed.

Reply:
I see, there are two visits which causes a rank mismatch. You can use `if constexpr` to check if the rank of the shapes match, something like: `if constexpr(update.get_shape().size() == output.get_shape().size())`. This can go before `gs_launch`.

Reply:
Thanks. I've made changes.